File: intellij-plugin/src/main/kotlin/dev/skillbill/intellij/infrastructure/cli/CliExecutableResolver.kt
package dev.skillbill.intellij.infrastructure.cli

import com.intellij.util.EnvironmentUtil
import dev.skillbill.intellij.application.PreferenceCachePort
import java.io.File
import java.nio.file.Files
import java.nio.file.Path

sealed class CliExecutableResolution {
    data class Found(val path: String, val source: CliExecutableSource) : CliExecutableResolution()
    data object Missing : CliExecutableResolution()
    data object Misconfigured : CliExecutableResolution()
}

enum class CliExecutableSource {
    OVERRIDE,
    SEARCH_PATH,
    INSTALL_DIRECTORY,
}

/**
 * Single lookup vocabulary for environment values, so resolution is testable without
 * an IDE and without mutating the JVM process environment.
 */
fun interface CliEnvironment {
    fun value(name: String): String?
}

/**
 * Resolves the `skill-bill` launcher for every CLI adapter.
 *
 * A desktop-launched IDE inherits the session environment, not the login shell's, so
 * `System.getenv("PATH")` alone misses the installer default `~/.local/bin` and every
 * other shell-profile entry. Lookup therefore merges the platform's shell-aware
 * environment with the process environment, then probes the installer's launcher
 * directory (`SKILL_BILL_BIN_DIR`, default `~/.local/bin`) before reporting the
 * executable missing.
 */
object CliExecutableResolver {
    const val EXECUTABLE_NAME: String = "skill-bill"

    private const val PATH_VARIABLE = "PATH"
    private const val BIN_DIR_VARIABLE = "SKILL_BILL_BIN_DIR"
    private const val HOME_VARIABLE = "HOME"

    fun resolve(
        preferences: PreferenceCachePort,
        environment: CliEnvironment = platformEnvironment(),
    ): CliExecutableResolution = resolveOverride(preferences.getCliExecutableOverride(), environment)

    fun resolveOverride(
        rawOverride: String?,
        environment: CliEnvironment = platformEnvironment(),
    ): CliExecutableResolution {
        val override = rawOverride?.trim()?.takeIf { it.isNotEmpty() }
        if (override != null) {
            val path = try {
                Path.of(override)
            } catch (_: Exception) {
                return CliExecutableResolution.Misconfigured
            }
            return if (isRunnable(path)) {
                CliExecutableResolution.Found(path.toString(), CliExecutableSource.OVERRIDE)
            } else {
                // Override set but unusable — do not fall back to PATH.
                CliExecutableResolution.Misconfigured
            }
        }
        findOnPath(EXECUTABLE_NAME, environment)?.let {
            return CliExecutableResolution.Found(it, CliExecutableSource.SEARCH_PATH)
        }
        findInInstallDirectories(EXECUTABLE_NAME, environment)?.let {
            return CliExecutableResolution.Found(it, CliExecutableSource.INSTALL_DIRECTORY)
        }
        return CliExecutableResolution.Missing
    }

    fun findOnPath(name: String, environment: CliEnvironment = platformEnvironment()): String? =
        splitSearchPath(environment.value(PATH_VARIABLE)).firstNotNullOfOrNull { directory ->
            runnableIn(directory, name)
        }

    fun installDirectories(environment: CliEnvironment = platformEnvironment()): List<String> = buildList {
        environment.value(BIN_DIR_VARIABLE)?.trim()?.takeIf { it.isNotEmpty() }?.let { add(it) }
        environment.value(HOME_VARIABLE)?.trim()?.takeIf { it.isNotEmpty() }?.let { home ->
            add(Path.of(home, ".local", "bin").toString())
        }
    }.distinct()

    /**
     * Merges the platform's shell-aware environment with the process environment instead of
     * choosing one: either can carry an entry the other lacks, and a missed entry is the
     * failure this resolver exists to prevent.
     */
    fun platformEnvironment(): CliEnvironment = CliEnvironment { name ->
        val shellValue = shellEnvironmentValue(name)
        val processValue = System.getenv(name) ?: fallbackProcessValue(name)
        if (name == PATH_VARIABLE) {
            mergeSearchPaths(shellValue, processValue)
        } else {
            shellValue?.takeIf { it.isNotBlank() } ?: processValue
        }
    }

    private fun findInInstallDirectories(name: String, environment: CliEnvironment): String? =
        installDirectories(environment).firstNotNullOfOrNull { directory -> runnableIn(directory, name) }

    private fun runnableIn(directory: String, name: String): String? {
        val candidate = try {
            Path.of(directory, name)
        } catch (_: Exception) {
            return null
        }
        return if (isRunnable(candidate)) candidate.toString() else null
    }

    private fun isRunnable(path: Path): Boolean = Files.isRegularFile(path) && Files.isExecutable(path)

    private fun shellEnvironmentValue(name: String): String? =
        try {
            EnvironmentUtil.getValue(name)
        } catch (_: Throwable) {
            null
        }

    private fun fallbackProcessValue(name: String): String? =
        if (name == HOME_VARIABLE) System.getProperty("user.home") else null

    private fun mergeSearchPaths(first: String?, second: String?): String? {
        val entries = (splitSearchPath(first) + splitSearchPath(second)).distinct()
        return entries.takeIf { it.isNotEmpty() }?.joinToString(File.pathSeparator)
    }

    private fun splitSearchPath(raw: String?): List<String> =
        raw?.split(File.pathSeparatorChar)
            ?.map { it.trim() }
            ?.filter { it.isNotEmpty() }
            ?: emptyList()
}

File: vscode-extension/src/infrastructure/cli/CliExecutableResolver.ts
import * as fs from "fs";
import * as os from "os";
import * as path from "path";
import { PreferenceCachePort } from "../../application/PreferenceCachePort";

export type CliExecutableResolution =
  | { kind: "found"; path: string; source: CliExecutableSource }
  | { kind: "missing" }
  | { kind: "misconfigured" };

export enum CliExecutableSource {
  OVERRIDE = "OVERRIDE",
  SEARCH_PATH = "SEARCH_PATH",
  INSTALL_DIRECTORY = "INSTALL_DIRECTORY",
}

export interface CliEnvironment {
  value(name: string): string | undefined;
}

export const EXECUTABLE_NAME = "skill-bill";

const PATH_VARIABLE = "PATH";
const BIN_DIR_VARIABLE = "SKILL_BILL_BIN_DIR";
const HOME_VARIABLE = "HOME";

export function resolveCliExecutable(
  preferences: PreferenceCachePort,
  environment: CliEnvironment = platformEnvironment(),
): CliExecutableResolution {
  return resolveOverride(preferences.getCliExecutableOverride(), environment);
}

export function resolveOverride(
  rawOverride: string | undefined,
  environment: CliEnvironment = platformEnvironment(),
): CliExecutableResolution {
  const override = rawOverride?.trim();
  if (override) {
    let candidate: string;
    try {
      candidate = path.resolve(override);
    } catch {
      return { kind: "misconfigured" };
    }
    if (isRunnable(candidate)) {
      return { kind: "found", path: candidate, source: CliExecutableSource.OVERRIDE };
    }
    return { kind: "misconfigured" };
  }
  const onPath = findOnPath(EXECUTABLE_NAME, environment);
  if (onPath) {
    return { kind: "found", path: onPath, source: CliExecutableSource.SEARCH_PATH };
  }
  const inInstall = findInInstallDirectories(EXECUTABLE_NAME, environment);
  if (inInstall) {
    return { kind: "found", path: inInstall, source: CliExecutableSource.INSTALL_DIRECTORY };
  }
  return { kind: "missing" };
}

export function findOnPath(name: string, environment: CliEnvironment = platformEnvironment()): string | undefined {
  for (const directory of splitSearchPath(environment.value(PATH_VARIABLE))) {
    const candidate = runnableIn(directory, name);
    if (candidate) {
      return candidate;
    }
  }
  return undefined;
}

export function installDirectories(environment: CliEnvironment = platformEnvironment()): string[] {
  const entries: string[] = [];
  const binDir = environment.value(BIN_DIR_VARIABLE)?.trim();
  if (binDir) {
    entries.push(binDir);
  }
  const home = environment.value(HOME_VARIABLE)?.trim();
  if (home) {
    entries.push(path.join(home, ".local", "bin"));
  }
  return [...new Set(entries)];
}

export function platformEnvironment(): CliEnvironment {
  return {
    value(name: string): string | undefined {
      if (name === PATH_VARIABLE) {
        return mergeSearchPaths(process.env.PATH, process.env.Path);
      }
      if (name === HOME_VARIABLE) {
        return process.env.HOME ?? process.env.USERPROFILE ?? os.homedir();
      }
      return process.env[name];
    },
  };
}

function findInInstallDirectories(name: string, environment: CliEnvironment): string | undefined {
  for (const directory of installDirectories(environment)) {
    const candidate = runnableIn(directory, name);
    if (candidate) {
      return candidate;
    }
  }
  return undefined;
}

function runnableIn(directory: string, name: string): string | undefined {
  let candidate: string;
  try {
    candidate = path.join(directory, name);
  } catch {
    return undefined;
  }
  return isRunnable(candidate) ? candidate : undefined;
}

function isRunnable(filePath: string): boolean {
  try {
    const stat = fs.statSync(filePath);
    if (!stat.isFile()) {
      return false;
    }
    if (process.platform === "win32") {
      return true;
    }
    fs.accessSync(filePath, fs.constants.X_OK);
    return true;
  } catch {
    return false;
  }
}

function mergeSearchPaths(...paths: (string | undefined)[]): string | undefined {
  const entries = paths.flatMap((raw) => splitSearchPath(raw));
  const distinct = [...new Set(entries)];
  return distinct.length > 0 ? distinct.join(path.delimiter) : undefined;
}

function splitSearchPath(raw: string | undefined): string[] {
  if (!raw) {
    return [];
  }
  return raw
    .split(path.delimiter)
    .map((entry) => entry.trim())
    .filter((entry) => entry.length > 0);
}
